dbms/src/Flash/Coprocessor/JoinInterpreterHelper.cpp

```diff
@@ -0,0 +1,332 @@
+            {
+                if (unlikely(join.probe_types(i).collate() != join.build_types(i).collate()))
+                    throw TiFlashException("Join with different collators on the join key", Errors::Coprocessor::BadRequest);
+                collators.push_back(getCollatorFromFieldType(join.probe_types(i)));
+            }
+            else
+                collators.push_back(nullptr);
+        }
+    return collators;
+}
+
+std::tuple<ExpressionActionsPtr, String, String> doGenJoinOtherConditionAction(
+    const Context & context,
+    const tipb::Join & join,
+    const NamesAndTypes & source_columns)
+{
+    if (join.other_conditions_size() == 0 && join.other_eq_conditions_from_in_size() == 0)
+        return {nullptr, "", ""};
+
+    DAGExpressionAnalyzer dag_analyzer(source_columns, context);
+    ExpressionActionsChain chain;
+
+    String filter_column_for_other_condition;
+    if (join.other_conditions_size() > 0)
+    {
+        std::vector<const tipb::Expr *> condition_vector;
+        for (const auto & c : join.other_conditions())
+        {
+            condition_vector.push_back(&c);
+        }
+        filter_column_for_other_condition = dag_analyzer.appendWhere(chain, condition_vector);
+    }
+
+    String filter_column_for_other_eq_condition;
+    if (join.other_eq_conditions_from_in_size() > 0)
+    {
+        std::vector<const tipb::Expr *> condition_vector;
+        for (const auto & c : join.other_eq_conditions_from_in())
+        {
+            condition_vector.push_back(&c);
+        }
+        filter_column_for_other_eq_condition = dag_analyzer.appendWhere(chain, condition_vector);
+    }
+
+    return {chain.getLastActions(), std::move(filter_column_for_other_condition), std::move(filter_column_for_other_eq_condition)};
+}
+} // namespace
+
+TiflashJoin::TiflashJoin(const tipb::Join & join_)
+    : join(join_)
+    , join_key_types(getJoinKeyTypes(join_))
+    , join_key_collators(getJoinKeyCollators(join_, join_key_types))
+{
+    std::tie(kind, build_side_index) = getJoinKindAndBuildSideIndex(join);
+    strictness = isSemiJoin() ? ASTTableJoin::Strictness::Any : ASTTableJoin::Strictness::All;
+}
+
+String TiflashJoin::genMatchHelperName(const Block & header1, const Block & header2) const
+{
+    if (!isLeftSemiFamily())
+    {
+        return "";
+    }
+
+    size_t i = 0;
+    String match_helper_name = fmt::format("{}{}", Join::match_helper_prefix, i);
+    while (header1.has(match_helper_name) || header2.has(match_helper_name))
+    {
+        match_helper_name = fmt::format("{}{}", Join::match_helper_prefix, ++i);
+    }
+    return match_helper_name;
+}
+
+NamesAndTypes TiflashJoin::genColumnsForOtherJoinFilter(
+    const Block & left_input_header,
+    const Block & right_input_header,
+    const ExpressionActionsPtr & prepare_join_actions1,
+    const ExpressionActionsPtr & prepare_join_actions2) const
+{
+    /// columns_for_other_join_filter is a vector of columns used
```

Comment:
It looks ok to me since it does not change the original logic, but as far as I can tell, just combine the output columns of `prepare_join_actions1` and `prepare_join_actions2` is enough. Please leave a todo here.

Reply:
done

Reply:
> It looks ok to me since it does not change the original logic, but as far as I can tell, just combine the output columns of `prepare_join_actions1` and `prepare_join_actions2` is enough. Please leave a todo here.

from https://github.com/pingcap/tiflash/blob/master/dbms/src/Flash/Coprocessor/DAGQueryBlockInterpreter.cpp#L313-L321
```
    /// Note the order in the column vector is very important:
    /// first the columns in left_input_header, then followed
    /// by the columns in right_input_header, if there are other
    /// columns generated before compile other join filter, then
    /// append the extra columns afterwards.
```
We cannot simply combine the output columns of `prepare_join_actions1` and `prepare_join_actions2`.

Reply:
These comments were wrote by me, but I can't remember why :(